### PR TITLE
heuristics to get title from section headers

### PR DESCRIPTION
**Pull request by Soeb-aryn**
If the first page of a document lacks a title, this heuristic identifies section headers or caption with the largest font size and promotes the most largest fontsize one to the title.

With the use_ocr flag enabled, the font size is determined based on the height of the bounding box. For Paddle OCR, each bounding box represents only a single line.
With the use_ocr flag enabled, the font size is obtained using PDFMiner.

### CODE REVIEW

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/detr_partitioner.py:132]: Replace with `if (font_size := m.properties.get("font_size"))` to reduce hash lookups

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/detr_partitioner.py:140]: Why do we need to filter if we only add if it is not None? We shouldn't need to do the conditional checks twice.

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/detr_partitioner.py:138]: Wouldn't `None` be better?

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/detr_partitioner.py:166]: Can you check this from the `text_extraction_options`? You'll also need to add documentation in the top level partition call to say that `promote_title` is an accepted argument in `text_extraction_options` and explain its functionality.

File: lib/sycamore/sycamore/transforms/partition.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/partition.py:458]: Move to `text_extraction_options`

File: lib/sycamore/sycamore/transforms/partition.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/partition.py:499]: As above.

File: lib/sycamore/sycamore/transforms/partition.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/partition.py:529]: As above.

File: lib/sycamore/sycamore/transforms/text_extraction/ocr_models.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/text_extraction/ocr_models.py:146]: `self.tesseract.get_text(image)` returns a tuple.

File: lib/sycamore/sycamore/transforms/text_extraction/ocr_models.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/text_extraction/ocr_models.py:116]: Can you add a comment that is not implemented? If it's not possible to calculate it from the output of tesseract, I would rather we return `None` to indicate this is not a valid font size

File: lib/sycamore/sycamore/transforms/text_extraction/pdf_miner.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/text_extraction/pdf_miner.py:92]: Can you rename this to `get_font_size` or something like it? `parse_obj` seems inaccurate for the functionality here.

File: lib/sycamore/sycamore/utils/pdf_utils.py
Comment (karanataryn) [inline lib/sycamore/sycamore/utils/pdf_utils.py:167]: Why don't we modify this to promote any set of elements to a title and take it in as an argument? You can set the default to be `["Section-header", "Caption"]` and call the function `promote_elements`. This would make it more widely applicable.

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (Soeb-aryn) [reply lib/sycamore/sycamore/transforms/detr_partitioner.py:166]: done

Comment (Soeb-aryn) [reply lib/sycamore/sycamore/transforms/detr_partitioner.py:166]: adding a 'output_label_options' option for driving all output label related functionalities.

File: lib/sycamore/sycamore/transforms/text_extraction/ocr_models.py
Comment (Soeb-aryn) [reply lib/sycamore/sycamore/transforms/text_extraction/ocr_models.py:116]: done

File: lib/sycamore/sycamore/utils/pdf_utils.py
Comment (Soeb-aryn) [reply lib/sycamore/sycamore/utils/pdf_utils.py:167]: good call, updating the function

File: lib/sycamore/sycamore/transforms/detr_partitioner.py
Comment (Soeb-aryn) [reply lib/sycamore/sycamore/transforms/detr_partitioner.py]: makes sense, shouldn't be needed here

**Review by karanataryn — COMMENTED** [review]
Looking better. We will need to add the documentation for this in this PR or a later one, I'll leave that to you.

File: lib/sycamore/sycamore/transforms/partition.py
Comment (karanataryn) [inline lib/sycamore/sycamore/transforms/partition.py:425]: I'm not sure if it makes sense to have `title_candidate_elements` as a top-level attribute given that we only care about it if `promote_title` is true but this is similar to `use_ocr` and `ocr_images` so it's not worth blocking on this.

**Review by karanataryn — APPROVED** [review]
Looks good to me. Could you add a small unit test to ensure we behave as intended? It should be relatively simple.

File: lib/sycamore/sycamore/tests/unit/utils/test_pdf_utils.py
Comment (karanataryn) [inline lib/sycamore/sycamore/tests/unit/utils/test_pdf_utils.py:187]: Need to remove this

File: lib/sycamore/sycamore/transforms/partition.py
Comment (MarkLindblad) [inline lib/sycamore/sycamore/transforms/partition.py:425]: The default specified in this docstring is wrong. The default for promote_title is False, not True.
https://github.com/aryn-ai/sycamore/blob/1f053475507aa249d3aa839e93e3a3f67027c3c8/lib/sycamore/sycamore/transforms/detr_partitioner.py#L205

Comment (MarkLindblad) [inline lib/sycamore/sycamore/transforms/partition.py:425]: Let's make it more clear that this is only done on the first page.

https://github.com/aryn-ai/sycamore/blob/59f8f2c95eb304365d6857fd756babdc30e22e88/lib/sycamore/sycamore/utils/pdf_utils.py#L171-L172

Comment (MarkLindblad) [reply lib/sycamore/sycamore/transforms/partition.py:425]: This has been fixed in #1057

Comment (MarkLindblad) [reply lib/sycamore/sycamore/transforms/partition.py:425]: This has been fixed in #1057

**Pull request merged**
